### DISCUSSION TITLE
Problem with testing and global state handling

### DISCUSSION THREAD

**Discussion by UniversalTraveller** [neuefische/questions#98] [Web] [answered]
Hello,

I wrote a function for testing user input with global state handling, like in this example:
[src/components/WithGlobalState/Form/Form.spec.js](https://github.com/aclaes/nf-nextjs-example-form-testing/blob/ae45ba7f8c388cc07a8f7af4a5e12a9762400ec7/src/components/WithGlobalState/Form/Form.spec.js)

When I run the tests, I get an error message:
`Matcher error: received value must be a mock or spy function`

What I tried was doing the test with a mock function like this:
`
const setSelectedEpisode = jest.fn;`

Than handing over the mock function when rendering the component:

`render(<DisplayEpisodes episodes={episodesProp onClick={setSelectedEpisode(episodes)} />);`

And testing whether the mock function has been called with the given prop:
`
expect(setSelectedEpisode).toHaveBeenCalledWith(episodes);`

This works, but it renders the test kind of useless, because it just tests the test.  How can I write a test that actually uses the function from my `useStore`, like in the exmaple?

My code is here:
https://github.com/UniversalTraveller/capstone-project/blob/a3a3bc79a35e86fc8aa1a802a5c139354cacab8e/src/components/DisplayEpisodes/DisplayEpisodes.spec.js

And that's the error message I'm getting:

```
FAIL  src/components/DisplayEpisodes/DisplayEpisodes.spec.js
  ● DisplayEpisodes component › should change the selected episode when the user clicks on an episode

    expect(received).toHaveBeenCalledTimes(expected)

    Matcher error: received value must be a mock or spy function

    Received has type:  function
    Received has value: [Function setSelectedEpisode]

      31 | 		await userEvent.click(selectEpisodeButton);
      32 |
    > 33 | 		expect(setSelectedEpisode).toHaveBeenCalledTimes(1);
         | 		                           ^
      34 | 		expect(setSelectedEpisode).toHaveBeenCalledWith(episodes);
      35 | 	});
      36 | 	it('should display the episodes title, date and length', () => {

      at Object.toHaveBeenCalledTimes (src/components/DisplayEpisodes/DisplayEpisodes.spec.js:33:30)
```

**Comment by alexclaes** [comment, marked as answer]
Hi Andreas,

it seems you are missing the mock for `zustand`.

It should look like this:
https://github.com/aclaes/nf-nextjs-example-form-testing/tree/main/src/__mocks__

You can copy the file's content into `src/__mocks__/zustand.js` and adjust `initialState` to the shape of your store. Please pay attention, that the file name is correct.

If you like to test `setSelectedEpisode`, this must be added as a mocked funktion.

```
initialState.setSelectedEpisode = jest.fn();
```

**Reply by UniversalTraveller** [reply]
Awesome, thank you Alex! I'm sure that's the bit I missed. Your link doesn't work, but i guess it's this file:
[src/__mocks__/zustand.js](https://github.com/aclaes/nf-nextjs-example-form-testing/blob/ae45ba7f8c388cc07a8f7af4a5e12a9762400ec7/src/__mocks__/zustand.js)

**Reply by alexclaes** [reply]
Yes! Sorry - I fixed the link
